actions/src/openstack_check_actions.py

```diff
@@ -0,0 +1,463 @@
+            logging.critical("The status code was: %s ", str(amphorae.status_code))
+            logging.critical("The JSON response was: \n %s", str(amph_json))
+            sys.exit(1)
+        return output
+
+    @staticmethod
+    def _check_amphora_status(amphorae):
+        try:
+            amph_json = amphorae.json()
+            return amph_json
+        except requests.exceptions.JSONDecodeError:
+            logging.critical(
+                msg="There was no JSON response \nThe status code was: "
+                + str(amphorae.status_code)
+                + "\nThe body was: \n"
+                + str(amphorae.content)
+            )
+            # pylint: disable=line-too-long
+            return (
+                "There was no JSON response \nThe status code was: "
+                + str(amphorae.status_code)
+                + "\nThe body was: \n"
+                + str(amphorae.content)
+            )
+
+    # pylint: disable=missing-function-docstring
+    @staticmethod
+    def _check_status(amphora):
+        # Extracts the status of the amphora and returns relevant info
+        status = amphora["status"]
+        if status in ("ALLOCATED", "BOOTING", "READY"):
+            return ["ok", status]
+
+        return ["error", status]
+
+    # pylint: disable=invalid-name
+    @staticmethod
+    def _ping_lb(ip):
+        # Runs the ping command to check that loadbalancer is up
+        response = os.system(
+            "ping -c 1 " + ip
+        )  # Might want to update to subprocess as this has been deprecated
+
+        # Checks output of ping command
+        if response == 0:
+            logging.info(msg="Successfully pinged " + ip)
+            return "success"
+
+        logging.info(msg=ip + " is down")
+        return "error"
+
+    def check_notify_snapshots(
+        self, cloud_account: str, project_id=None, all_projects=False
+    ):
+        # pylint: disable=line-too-long
+        output = {
+            "title": "Project {p[name]} has an old volume snapshot",
+            "body": "The volume snapshot was last updated on: {p[updated]}\nSnapshot name: {p[name]}\nSnapshot id: {p[id]}\nProject id: {p[project_id]}",
+            "server_list": [],
+        }
+        snapshots = []
+        if all_projects:
+            with OpenstackConnection(cloud_name=cloud_account) as conn:
+                projects = conn.list_projects()
+            for project in projects:
+                snapshots.extend(
+                    self.check_snapshots(
+                        project=project["id"], cloud_account=cloud_account
+                    )
+                )
+        else:
+            snapshots = self.check_snapshots(
+                project=project_id, cloud_account=cloud_account
+            )
+
+        for snapshot in snapshots:
+            with OpenstackConnection(cloud_name=cloud_account) as conn:
+                proj = conn.get_project(name_or_id=snapshot["project_id"])
```

Comment:
Could we swap these, open the connection once then do a loop within it.
This way we don't hammer the auth server constantly opening new connections each time